File: Tennis/TennisGame1.cs
using System;

namespace Tennis
{
	class TennisGame1 : ITennisGame
	{
		private readonly TennisGameState _tennisGameState = new TennisGameState();

		public void WonPoint(string playerName)
		{
			if (playerName == "player1")
				_tennisGameState.Player1Score += 1;
			else
				_tennisGameState.Player2Score += 1;
		}

		public string GetScore()
		{
			if (_tennisGameState.Player1Score == _tennisGameState.Player2Score)
			{
				if (_tennisGameState.Player1Score < 3)
				{
					return ConvertScoreToString(_tennisGameState.Player1Score) + "-" + "All";
				}

				return "Deuce";
			}
			if (_tennisGameState.Player1Score >= 4 || _tennisGameState.Player2Score >= 4)
			{
				var minusResult = _tennisGameState.Player1Score - _tennisGameState.Player2Score;
				if (minusResult == 1)
				{
					return "Advantage player1";
				}
				if (minusResult == -1)
				{
					return "Advantage player2";
				}
				if (minusResult >= 2)
				{
					return "Win for player1";
				}
				return "Win for player2";
			}
			return ConvertScoreToString(_tennisGameState.Player1Score) + "-" + ConvertScoreToString(_tennisGameState.Player2Score);
		}

		private static string ConvertScoreToString(int score)
		{
			switch (score)
			{
				case 0:
					return "Love";
				case 1:
					return "Fifteen";
				case 2:
					return "Thirty";
				case 3:
					return "Forty";
			}
			throw new ArgumentOutOfRangeException(nameof(score));
		}
	}
}

File: TennisFSharp/TennisGame1.fs
namespace Tennis

type ITennisGame =
    abstract WonPoint : string -> ITennisGame
    abstract GetScore : unit -> string

type TennisGameState = {player1Score: int; player2Score: int}

type TennisGame1(?player1Score: int, ?player2Score: int) =
    let orZero value = defaultArg value 0
    let _state:TennisGameState = 
        {player1Score = player1Score |> orZero;
         player2Score = player2Score |> orZero}
    let ConvertScoreToString(score) =
        match score with
        | 0 -> Some("Love")
        | 1 -> Some("Fifteen")
        | 2 -> Some("Thirty")
        | 3 -> Some("Forty")
        | _ -> None
    let GetScoreIfTieDuringRegularPlay({player1Score = player1Score; player2Score = player2Score}) =
        if player1Score = player2Score && player1Score < 3 then
            Some(ConvertScoreToString(player1Score).Value + "-" + "All")
        else
            None
    let GetScoreIfTieDuringExtendedPlay({player1Score = player1Score; player2Score = player2Score}) =
        if player1Score = player2Score && player1Score >= 3 then
            Some("Deuce")
        else
            None
    let GetScoreIfWinOrAdvantage({player1Score = player1Score; player2Score = player2Score}) =
        if player1Score > 3 || player2Score > 3 then
            let diff = player1Score - player2Score
            match diff with
            | 1 -> Some("Advantage player1")
            | -1 -> Some("Advantage player2")
            | diff when diff >= 2 -> Some("Win for player1")
            | diff when diff <= -2 -> Some("Win for player2")
            | _ -> None
        else
            None
    let GetScoreForNormalPlayWhenNotATie({player1Score = player1Score; player2Score = player2Score}) =
        if player1Score <> player2Score && player1Score <= 3 then
            Some(ConvertScoreToString(player1Score).Value + "-" + ConvertScoreToString(player2Score).Value)
        else
            None
    let ChainStateIfNone (ruleFunc: (TennisGameState -> Option<string>)) (maybeResult: Option<string>) =
        match maybeResult with
        | Some result -> Some(result)
        | None -> ruleFunc(_state)
    interface ITennisGame with
        member this.WonPoint(playerName) =
            match playerName with
            | "player1" -> TennisGame1(_state.player1Score + 1, _state.player2Score) :> ITennisGame
            | _ -> TennisGame1(_state.player1Score, _state.player2Score + 1) :> ITennisGame
        member this.GetScore() =
            GetScoreIfTieDuringRegularPlay(_state)
                |> ChainStateIfNone GetScoreIfTieDuringExtendedPlay
                |> ChainStateIfNone GetScoreIfWinOrAdvantage
                |> ChainStateIfNone GetScoreForNormalPlayWhenNotATie
                |> Option.get
